imap_processing/tests/cdf/test_utils.py

```diff
@@ -1,21 +1,26 @@
+"""Tests for the ``cdf.utils`` module."""
+
 import imap_data_access
 import numpy as np
+import pytest
 import xarray as xr
 
 from imap_processing import launch_time
 from imap_processing.cdf.global_attrs import ConstantCoordinates
-from imap_processing.cdf.utils import calc_start_time, write_cdf
+from imap_processing.cdf.utils import calc_start_time, load_cdf, write_cdf
 from imap_processing.swe.swe_cdf_attrs import swe_l1a_global_attrs
 
 
-def test_calc_start_time():
-    assert calc_start_time(0) == launch_time
-    assert calc_start_time(1) == launch_time + np.timedelta64(1, "s")
+@pytest.fixture()
+def test_dataset():
```

Comment:
I think I follow what you are saying here but just want to be clear. Are you talking about adding a test like this?

```python
def test_written_and_loaded_dataset(test_dataset):

    assert test_dataset == load_cdf(write_cdf(test_dataset))
```

to make sure that nothing i getting changed in the `xarray` object when passed through the `write_cdf()` and `load_cdf()` function?

Reply:
Yep, exactly! Making sure it is a reversible operation and we are getting the same data in and out.

Reply:
Gotcha. Unfortunately it is indeed breaking, so good that we checked!

Reply:
Uh oh! Hopefully not too big of an issue? Does it need to be fixed in cdflib or on our end... there is hardly anything we are doing here in these functions.

Reply:
It is kind of hard to untangle, and it is looking like the changes are coming from the `cdflib` side. For example:

```
<xarray.Dataset> Size: 24B
Dimensions:  (record0: 3)
Dimensions without coordinates: record0
Data variables:
    epoch    (record0) datetime64[ns] 24B 2010-01-01T00:01:01 ... 2010-01-01T...
```

gets changed to

```
<xarray.Dataset> Size: 24B
Dimensions:  (epoch: 3)
Coordinates:
  * epoch    (epoch) datetime64[ns] 24B 2010-01-01T00:01:01 ... 2010-01-01T00...
Data variables:
    *empty*
```

Reply:
Yes, sorry, I had those flipped. So the `test_dataset` is (correctly) being defined as

```
<xarray.Dataset> Size: 24B
Dimensions:  (epoch: 3)
Coordinates:
  * epoch    (epoch) datetime64[ns] 24B 2010-01-01T00:01:01 ... 2010-01-01T00...
Data variables:
    *empty*
```

but after writing and loading, the coordinates are removed and "epoch" becomes a data variable:

```
Dimensions:  (record0: 3)
Dimensions without coordinates: record0
Data variables:
    epoch    (record0) datetime64[ns] 24B 2010-01-01T00:01:01 ... 2010-01-01T...

```

Reply:
I think I have identified that the transformation is happening within `cdflib`:

```python
In [1]: import cdflib
   ...: from cdflib import xarray
   ...: import xarray as xr
   ...: import numpy as np
   ...: from imap_processing.cdf.global_attrs import ConstantCoordinates
   ...: from imap_processing.swe.swe_cdf_attrs import swe_l1a_global_attrs

In [2]: filename = 'imap_swe_l1_sci_20100101_v001.cdf'

   ...:                 [
   ...:                     np.datetime64("2010-01-01T00:01:01", "ns"),
   ...:                     np.datetime64("2010-01-01T00:01:02", "ns"),
   ...:                     np.datetime64("2010-01-01T00:01:03", "ns"),
   ...:                 ],
   ...:             ),
   ...:         },
   ...:         attrs=swe_l1a_global_attrs.output()
   ...:         | {"Logical_source": "imap_swe_l1_sci", "Data_version": "001"},
   ...:     )
   ...:     dataset["epoch"].attrs = ConstantCoordinates.EPOCH

In [4]: cdflib.xarray.xarray_to_cdf(dataset, filename)

In [5]: new_dataset = cdflib.xarray.cdf_to_xarray(filename)

In [6]: new_dataset
Out[6]: 
<xarray.Dataset> Size: 24B
Dimensions:  (record0: 3)
Dimensions without coordinates: record0
Data variables:
    epoch    (record0) int64 24B 1262304061000000000 ... 1262304063000000000
Attributes: (12/15)
    Project:                     ['STP>Solar-Terrestrial Physics']
    Source_name:                 ['IMAP>Interstellar Mapping and Acceleration...
    Discipline:                  ['Solar Physics>Heliospheric Physics']
    Mission_group:               ['IMAP>Interstellar Mapping and Acceleration...
    PI_name:                     ['Dr. David J. McComas']
    PI_affiliation:              ['Princeton Plasma Physics Laboratory', '100...
    ...                          ...
    TEXT:                        ['The Solar Wind Electron (SWE) instrument m...
    Instrument_type:             ['Particles (space)']
    Logical_file_id:             ['FILL ME IN AT FILE CREATION']
    Data_type:                   ['L1A_SCI>Level-1A Science Data']
    Logical_source:              ['imap_swe_l1_sci']
    Logical_source_description:  ['IMAP Mission SWE Instrument Level-1A Data']

```

Reply:
ping @bryan-harter I think this is a bug within cdflib, but maybe it isn't so much a bug as a limitation of CDF? Is there any way to keep the coordinates when loading the cdf from disk into an xarray dataset?

Reply:
So the way cdf_to_xarray tells what the dimensions are is only through what the other variables in the file are pointing at (i.e. with a DEPEND_0/1/2/3/etc).  

If there are no other variables in the file, then cdf_to_xarray doesn't consider "epoch" a dimension. So its more of a limitation of CDF I suppose.  One way we can get around this is to add a DEPEND_0 to the epoch variable that just points to itself.  So if you add the following, the code will identify the dimension correctly:

`dataset["epoch"].attrs['DEPEND_0'] = 'epoch'`

Alternatively you can just add some dummy variable, like: 

`dataset = xr.Dataset(
         {
             "epoch": (
                 "epoch",
                 [
                     np.datetime64("2010-01-01T00:01:01", "ns"),
                     np.datetime64("2010-01-01T00:01:02", "ns"),
                     np.datetime64("2010-01-01T00:01:03", "ns"),
                 ],
             ),
             "hello": (
                 "hello",
                 [
                     1,
                     2,
                     3,
                 ],
             ),
         },
         attrs=swe_l1a_global_attrs.output()
         | {"Logical_source": "imap_swe_l1_sci", "Data_version": "001"},
     )
dataset["epoch"].attrs = ConstantCoordinates.EPOCH
dataset["hello"].attrs = swe_metadata_attrs.output()
`

In this case, since the variable "hello" is points to "epoch", then cdf_to_xarray correctly identifies "epoch" as a dimension

Reply:
I guess on the cdflib side, I can try to make changes so that 1D "support_variables" are always their own dimension? I can try taking a look at that and see if that fundamentally breaks anything.  But it handles time variables slightly differently than the other DEPENDs.

Reply:
@bryan-harter Thanks for the info on that. Your simple fix of adding the `DEPEND_0` appears to be working as intended!

One other thing that seems to be changing is that `cdflib` is writing and/or loading the CDF attributes into lists, i.e.:

```
Attributes: (12/15)
    Project:                     STP>Solar-Terrestrial Physics
    Source_name:                 IMAP>Interstellar Mapping and Acceleration P...
    Discipline:                  Solar Physics>Heliospheric Physics
    Mission_group:               IMAP>Interstellar Mapping and Acceleration P...
    PI_name:                     Dr. David J. McComas
    PI_affiliation:              ('Princeton Plasma Physics Laboratory', '100...
    ...                          ...
    TEXT:                        The Solar Wind Electron (SWE) instrument mea...
    Instrument_type:             Particles (space)
    Logical_file_id:             ['FILL ME IN AT FILE CREATION']
    Data_type:                   L1A_SCI>Level-1A Science Data
    Logical_source:              imap_swe_l1_sci
    Logical_source_description:  IMAP Mission SWE Instrument Level-1A Data
```

gets turned into

```
Attributes: (12/15)
    Project:                     ['STP>Solar-Terrestrial Physics']
    Source_name:                 ['IMAP>Interstellar Mapping and Acceleration...
    Discipline:                  ['Solar Physics>Heliospheric Physics']
    Mission_group:               ['IMAP>Interstellar Mapping and Acceleration...
    PI_name:                     ['Dr. David J. McComas']
    PI_affiliation:              ['Princeton Plasma Physics Laboratory', '100...
    ...                          ...
    TEXT:                        ['The Solar Wind Electron (SWE) instrument m...
    Instrument_type:             ['Particles (space)']
    Logical_file_id:             ['FILL ME IN AT FILE CREATION']
    Data_type:                   ['L1A_SCI>Level-1A Science Data']
    Logical_source:              ['imap_swe_l1_sci']
    Logical_source_description:  ['IMAP Mission SWE Instrument Level-1A Data']
```

Do you know if this is expected / purposeful?